File: tmp/helengine-core-cpp-regenerated/ScrollComponent.cpp
#ifdef DrawText
#undef DrawText
#endif
#include "ScrollComponent.hpp"
#include "float3.hpp"
#include "Entity.hpp"
#include "Core.hpp"
#include "InputSystem.hpp"
#include "runtime/array.hpp"
#include "runtime/finally.hpp"
#include "runtime/native_cast.hpp"
#include "runtime/native_datetime.hpp"
#include "runtime/native_dictionary.hpp"
#include "runtime/native_disposable.hpp"
#include "runtime/native_enum.hpp"
#include "runtime/native_event.hpp"
#include "runtime/native_exceptions.hpp"
#include "runtime/native_list.hpp"
#include "runtime/native_nullable.hpp"
#include "runtime/native_span.hpp"
#include "runtime/native_stack.hpp"
#include "runtime/native_string.hpp"
#include "runtime/native_tuple.hpp"
#include "runtime/native_type.hpp"
#include "system/app_context.hpp"
#include "system/binary_primitives.hpp"
#include "system/bit_converter.hpp"
#include "system/diagnostics/debug.hpp"
#include "system/diagnostics/stopwatch.hpp"
#include "system/guid.hpp"
#include "system/io/file-stream.hpp"
#include "system/io/file.hpp"
#include "system/io/memory-stream.hpp"
#include "system/io/path.hpp"
#include "system/io/stream-reader.hpp"
#include "system/io/stream.hpp"
#include "system/io/string-reader.hpp"
#include "system/math.hpp"
#include "system/number.hpp"
#include "system/security/cryptography/sha256.hpp"
#include "system/string_comparer.hpp"
#include "system/text/encoding.hpp"
#include "system/text/regular_expressions/regex.hpp"
#include "system/text/string-builder.hpp"

ScrollComponent::ScrollComponent() : ScrollOffset(0), ScrollOffsetChanged(), ItemCountValue(0), RequiresPointerInsideValue(true), ScrollStepCountValue(1), SizeValue(), VisibleItemCountValue(0), WheelNotchSizeValue(StandardWheelNotch)
{
}

int32_t ScrollComponent::get_ItemCount()
{
return this->ItemCountValue;}

void ScrollComponent::set_ItemCount(int32_t value)
{
    if (value < 0)
    {
throw ([&]() {
auto __ctor_arg_0000019A = "value";
auto __ctor_arg_0000019B = "Item count must be zero or greater.";
return new ArgumentOutOfRangeException(__ctor_arg_0000019A, __ctor_arg_0000019B);
})();
    }
this->ItemCountValue = value;
this->ClampScrollOffset();
}

int32_t ScrollComponent::get_MaximumScrollOffset()
{
return Math::Max(0, this->ItemCountValue - this->VisibleItemCountValue);}

bool ScrollComponent::get_RequiresPointerInside()
{
return this->RequiresPointerInsideValue;}

void ScrollComponent::set_RequiresPointerInside(bool value)
{
this->RequiresPointerInsideValue = value;
}

int32_t ScrollComponent::get_ScrollOffset()
{
return this->ScrollOffset;
}

void ScrollComponent::set_ScrollOffset(int32_t value)
{
this->ScrollOffset = value;
}

int32_t ScrollComponent::get_ScrollStepCount()
{
return this->ScrollStepCountValue;}

void ScrollComponent::set_ScrollStepCount(int32_t value)
{
    if (value < 1)
    {
throw ([&]() {
auto __ctor_arg_0000019C = "value";
auto __ctor_arg_0000019D = "Scroll step count must be at least one.";
return new ArgumentOutOfRangeException(__ctor_arg_0000019C, __ctor_arg_0000019D);
})();
    }
this->ScrollStepCountValue = value;
}

int2* ScrollComponent::get_Size()
{
return this->SizeValue;}

void ScrollComponent::set_Size(int2* value)
{
    if (value->X < 0 || value->Y < 0)
    {
throw ([&]() {
auto __ctor_arg_0000019E = "value";
auto __ctor_arg_0000019F = "Scroll viewport size must not be negative.";
return new ArgumentOutOfRangeException(__ctor_arg_0000019E, __ctor_arg_0000019F);
})();
    }
this->SizeValue = value;
}

int32_t ScrollComponent::get_VisibleItemCount()
{
return this->VisibleItemCountValue;}

void ScrollComponent::set_VisibleItemCount(int32_t value)
{
    if (value < 0)
    {
throw ([&]() {
auto __ctor_arg_000001A0 = "value";
auto __ctor_arg_000001A1 = "Visible item count must be zero or greater.";
return new ArgumentOutOfRangeException(__ctor_arg_000001A0, __ctor_arg_000001A1);
})();
    }
this->VisibleItemCountValue = value;
this->ClampScrollOffset();
}

int32_t ScrollComponent::get_WheelNotchSize()
{
return this->WheelNotchSizeValue;}

void ScrollComponent::set_WheelNotchSize(int32_t value)
{
    if (value < 1)
    {
throw ([&]() {
auto __ctor_arg_000001A2 = "value";
auto __ctor_arg_000001A3 = "Wheel notch size must be at least one.";
return new ArgumentOutOfRangeException(__ctor_arg_000001A2, __ctor_arg_000001A3);
})();
    }
this->WheelNotchSizeValue = value;
}

void ScrollComponent::ClampScrollOffset()
{
this->set_ScrollOffset(this->ClampOffset(this->ScrollOffset));
}

bool ScrollComponent::ContainsScreenPoint(int32_t x, int32_t y)
{
    if (Parent == nullptr)
    {
return false;    }
::float3 origin = Parent->get_Position();
return x >= origin.X && x < origin.X + this->SizeValue->X && y >= origin.Y && y < origin.Y + this->SizeValue->Y;}

void ScrollComponent::ResetScrollOffset()
{
this->set_ScrollOffset(0);
}

bool ScrollComponent::ScrollTo(int32_t scrollOffset)
{
return this->SetScrollOffset(scrollOffset, true);}

bool ScrollComponent::TryApplyWheelInput()
{
    if (Parent == nullptr)
    {
return false;    }
    if (this->get_MaximumScrollOffset() <= 0)
    {
return false;    }
    if (this->RequiresPointerInsideValue && !this->ContainsScreenPoint(Core::get_Instance()->get_Input()->GetMouseX(), Core::get_Instance()->get_Input()->GetMouseY()))
    {
return false;    }
const int32_t wheelDelta = Core::get_Instance()->get_Input()->GetMouseScrollWheelDelta();
    if (wheelDelta == 0)
    {
return false;    }
int32_t scrollSteps = wheelDelta / this->WheelNotchSizeValue;
    if (scrollSteps == 0)
    {
scrollSteps = wheelDelta > 0 ? 1 : -1;
    }
scrollSteps *= this->ScrollStepCountValue;
const int32_t nextOffset = this->ScrollOffset - scrollSteps;
return this->SetScrollOffset(nextOffset, true);}

void ScrollComponent::Update()
{
this->TryApplyWheelInput();
}

uint8_t ScrollComponent::get_UpdateOrder()
{
return this->UpdateComponent::get_UpdateOrder();
}

void ScrollComponent::set_UpdateOrder(uint8_t value)
{
this->UpdateComponent::set_UpdateOrder(value);
}

::Entity* ScrollComponent::get_Parent()
{
return this->Component::get_Parent();
}

void ScrollComponent::set_Parent(::Entity* value)
{
this->Component::set_Parent(value);
}

int32_t ScrollComponent::StandardWheelNotch = 120;

int32_t ScrollComponent::ClampOffset(int32_t scrollOffset)
{
const int32_t maxOffset = this->get_MaximumScrollOffset();
    if (scrollOffset < 0)
    {
return 0;    }
    if (scrollOffset > maxOffset)
    {
return maxOffset;    }
return scrollOffset;}

bool ScrollComponent::SetScrollOffset(int32_t scrollOffset, bool raiseEvent)
{
const int32_t clampedOffset = this->ClampOffset(scrollOffset);
    if (clampedOffset == this->ScrollOffset)
    {
return false;    }
this->set_ScrollOffset(clampedOffset);
    if (raiseEvent && true)
    {
this->ScrollOffsetChanged.Invoke(this, ScrollOffset);
    }
return true;}

File: engine/helengine.core/components/2d/interactable/ScrollComponent.cs
namespace helengine {
    /// <summary>
    /// Tracks wheel-driven list scrolling for a rectangular viewport, acts as its own clip region, derives the visible range when needed, and can translate a bound content root automatically.
    /// </summary>
    public class ScrollComponent : UpdateComponent, IClipRegion2D {
        /// <summary>
        /// Standard mouse-wheel delta used to represent one notch on Windows-compatible devices.
        /// </summary>
        const int StandardWheelNotch = 120;

        /// <summary>
        /// Size of the scroll viewport in screen-space pixels.
        /// </summary>
        int2 SizeValue;

        /// <summary>
        /// Number of items that can appear in the backing list.
        /// </summary>
        int ItemCountValue;

        /// <summary>
        /// Pixel height or extent consumed by one item in the scrolling content.
        /// </summary>
        int ItemExtentValue = 1;

        /// <summary>
        /// Number of items visible inside the current viewport, or zero when the component should derive it automatically.
        /// </summary>
        int VisibleItemCountValue;

        /// <summary>
        /// Number of items to move for each wheel notch.
        /// </summary>
        int ScrollStepCountValue = 1;

        /// <summary>
        /// Wheel delta that maps to one scroll notch.
        /// </summary>
        int WheelNotchSizeValue = StandardWheelNotch;

        /// <summary>
        /// Tracks whether wheel scrolling should only occur while the pointer is inside the viewport.
        /// </summary>
        bool RequiresPointerInsideValue = true;

        /// <summary>
        /// Optional content root that should be translated when the scroll offset changes.
        /// </summary>
        Entity ContentRootValue;

        /// <summary>
        /// Raised when the scroll offset changes because of wheel input or an explicit scroll request.
        /// </summary>
        public event Action<ScrollComponent, int> ScrollOffsetChanged;

        /// <summary>
        /// Gets or sets the viewport size used for pointer hit testing.
        /// </summary>
        public int2 Size {
            get { return SizeValue; }
            set {
                if (value.X < 0 || value.Y < 0) {
                    throw new ArgumentOutOfRangeException(nameof(value), "Scroll viewport size must not be negative.");
                }

                SizeValue = value;
            }
        }

        /// <summary>
        /// Gets or sets the total number of list items that can be scrolled through.
        /// </summary>
        public int ItemCount {
            get { return ItemCountValue; }
            set {
                if (value < 0) {
                    throw new ArgumentOutOfRangeException(nameof(value), "Item count must be zero or greater.");
                }

                ItemCountValue = value;
                ClampScrollOffset();
            }
        }

        /// <summary>
        /// Gets or sets the number of items visible in the viewport.
        /// When set to zero, the component derives the value from the viewport height and item extent.
        /// </summary>
        public int VisibleItemCount {
            get { return GetVisibleItemCount(); }
            set {
                if (value < 0) {
                    throw new ArgumentOutOfRangeException(nameof(value), "Visible item count must be zero or greater.");
                }

                VisibleItemCountValue = value;
                ClampScrollOffset();
            }
        }

        /// <summary>
        /// Gets or sets the item extent used when the visible item count is derived automatically.
        /// </summary>
        public int ItemExtent {
            get { return ItemExtentValue; }
            set {
                if (value < 1) {
                    throw new ArgumentOutOfRangeException(nameof(value), "Item extent must be at least one.");
                }

                ItemExtentValue = value;
                ClampScrollOffset();
                ApplyContentRootOffset();
            }
        }

        /// <summary>
        /// Gets the maximum scroll offset allowed by the current item and viewport counts.
        /// </summary>
        public int MaximumScrollOffset {
            get { return Math.Max(0, ItemCountValue - GetVisibleItemCount()); }
        }

        /// <summary>
        /// Gets the current scroll offset in item units.
        /// </summary>
        public int ScrollOffset { get; private set; }

        /// <summary>
        /// Gets or sets the content root that should move in response to the current scroll offset.
        /// </summary>
        public Entity ContentRoot {
            get { return ContentRootValue; }
            set {
                ContentRootValue = value;
                ApplyContentRootOffset();
            }
        }

        /// <summary>
        /// Gets or sets the number of items to move for each wheel notch.
        /// </summary>
        public int ScrollStepCount {
            get { return ScrollStepCountValue; }
            set {
                if (value < 1) {
                    throw new ArgumentOutOfRangeException(nameof(value), "Scroll step count must be at least one.");
                }

                ScrollStepCountValue = value;
            }
        }

        /// <summary>
        /// Gets or sets the wheel delta that corresponds to one scroll notch.
        /// </summary>
        public int WheelNotchSize {
            get { return WheelNotchSizeValue; }
            set {
                if (value < 1) {
                    throw new ArgumentOutOfRangeException(nameof(value), "Wheel notch size must be at least one.");
                }

                WheelNotchSizeValue = value;
            }
        }

        /// <summary>
        /// Gets or sets a value indicating whether the pointer must be inside the viewport before scrolling can occur.
        /// </summary>
        public bool RequiresPointerInside {
            get { return RequiresPointerInsideValue; }
            set { RequiresPointerInsideValue = value; }
        }

        /// <summary>
        /// Advances scrolling from the active mouse wheel while the pointer is inside the viewport.
        /// </summary>
        public override void Update() {
            TryApplyWheelInput();
        }

        /// <summary>
        /// Returns true when the provided screen point lies inside the current viewport bounds.
        /// </summary>
        /// <param name="x">Pointer X coordinate in screen coordinates.</param>
        /// <param name="y">Pointer Y coordinate in screen coordinates.</param>
        /// <returns>True when the point lies inside the scroll viewport.</returns>
        public bool ContainsScreenPoint(int x, int y) {
            if (Parent == null) {
                return false;
            }

            float4 viewportRect = GetClipRect();
            return x >= viewportRect.X &&
                   x < viewportRect.X + viewportRect.Z &&
                   y >= viewportRect.Y &&
                   y < viewportRect.Y + viewportRect.W;
        }

        /// <summary>
        /// Resets the scroll offset to the first visible item without raising a change event.
        /// </summary>
        public void ResetScrollOffset() {
            SetScrollOffset(0, false);
        }

        /// <summary>
        /// Clamps the current scroll offset to the active item range without raising a change event.
        /// </summary>
        public void ClampScrollOffset() {
            SetScrollOffset(ScrollOffset, false);
        }

        /// <summary>
        /// Applies one explicit scroll offset and notifies listeners when the value changes.
        /// </summary>
        /// <param name="scrollOffset">Desired scroll offset in item units.</param>
        /// <returns>True when the offset changed.</returns>
        public bool ScrollTo(int scrollOffset) {
            return SetScrollOffset(scrollOffset, true);
        }

        /// <summary>
        /// Attempts to apply wheel input from the current frame.
        /// </summary>
        /// <returns>True when the scroll offset changed.</returns>
        public bool TryApplyWheelInput() {
            if (Parent == null) {
                return false;
            }

            if (MaximumScrollOffset <= 0) {
                return false;
            }

            if (RequiresPointerInsideValue && !ContainsScreenPoint(Core.Instance.Input.GetMouseX(), Core.Instance.Input.GetMouseY())) {
                return false;
            }

            int wheelDelta = Core.Instance.Input.GetMouseScrollWheelDelta();
            if (wheelDelta == 0) {
                return false;
            }

            int scrollSteps = wheelDelta / WheelNotchSizeValue;
            if (scrollSteps == 0) {
                scrollSteps = wheelDelta > 0 ? 1 : -1;
            }

            scrollSteps *= ScrollStepCountValue;
            int nextOffset = ScrollOffset - scrollSteps;
            return SetScrollOffset(nextOffset, true);
        }

        /// <summary>
        /// Applies one scroll offset and optionally raises the change event.
        /// </summary>
        /// <param name="scrollOffset">Requested scroll offset.</param>
        /// <param name="raiseEvent">True to raise the offset changed event.</param>
        /// <returns>True when the offset changed.</returns>
        bool SetScrollOffset(int scrollOffset, bool raiseEvent) {
            int clampedOffset = ClampOffset(scrollOffset);
            if (clampedOffset == ScrollOffset) {
                return false;
            }

            ScrollOffset = clampedOffset;
            if (raiseEvent && ScrollOffsetChanged != null) {
                ScrollOffsetChanged(this, ScrollOffset);
            }

            ApplyContentRootOffset();
            return true;
        }

        /// <summary>
        /// Applies the current scroll offset to the bound content root when one exists.
        /// </summary>
        void ApplyContentRootOffset() {
            if (ContentRootValue == null) {
                return;
            }

            float3 position = ContentRootValue.LocalPosition;
            ContentRootValue.LocalPosition = new float3(position.X, -(ScrollOffset * ItemExtentValue), position.Z);
        }

        /// <summary>
        /// Gets the clip rectangle used by descendants and pointer hit testing.
        /// </summary>
        /// <returns>Viewport rectangle expressed as X, Y, Width, Height.</returns>
        public float4 GetClipRect() {
            if (Parent == null) {
                throw new InvalidOperationException("Scroll components require an attached parent entity.");
            }

            float3 origin = Parent.Position;
            return new float4(origin.X, origin.Y, SizeValue.X, SizeValue.Y);
        }

        /// <summary>
        /// Clamps one requested scroll offset to the current available range.
        /// </summary>
        /// <param name="scrollOffset">Requested offset.</param>
        /// <returns>Clamped offset value.</returns>
        int ClampOffset(int scrollOffset) {
            int maxOffset = MaximumScrollOffset;
            if (scrollOffset < 0) {
                return 0;
            }

            if (scrollOffset > maxOffset) {
                return maxOffset;
            }

            return scrollOffset;
        }

        /// <summary>
        /// Resolves the item count visible in the current viewport.
        /// </summary>
        /// <returns>Visible item count derived from the viewport or an explicit override.</returns>
        int GetVisibleItemCount() {
            if (VisibleItemCountValue > 0) {
                return VisibleItemCountValue;
            }

            int extent = ItemExtentValue;
            if (extent <= 0) {
                return 1;
            }

            return Math.Max(1, (SizeValue.Y + extent - 1) / extent);
        }
    }
}

